### ISSUE TITLE
Provide a way to specify a subprotocol in client-server

### ISSUE THREAD

**Issue by vasilvv** [w3c/webrtc-quic#110]
WebSockets have a subprotocol mechanism which works similar to ALPN (see https://tools.ietf.org/html/rfc6455#section-1.9).  Perhaps we should provide web applications using this API the same functionality directly via ALPN, e.g., if the client asks for "foo" and "bar", we provide ALPNs "wq:foo" and "wq:bar".

**Comment by vasilvv** [comment]
> Would this make sense as a transport parameter?

I actually wanted to just use the TLS ALPN fields for that (with a namespace prefix, of course, so that you can't negotiate non-web ALPNs with this).

**Comment by pthatcherg** [comment]
Oh, sorry, I didn't understand your first comment. 

So basically you want to stuff extra things in the ALPN.  Does that mean that the server sends back an ALPN value to the client (a response to the given ALPN)?  And the client can send more than one ALPN value?  

Sorry for my ignorance of ALPNs :).

**Comment by aboba** [comment]
How does protocol multiplexing fit into the ALPN discussion?  ALPN usage is clear if an RTCQuicTransport is only used for one purpose (e.g. CSQUIC, HTTP, etc.) but if we are to allow "pooled connections" where an RTCQuicTransport is used for HTTP *and* CSQUIC, then how what ALPN is used and does this open security holes?

**Comment by vasilvv** [comment]
ALPN works the following way.  Client sends a list of protocols that it supports, and the server picks one of them.  In QUIC, if the client and the server don't have a protocol in common, it's a fatal error.

So, for example, the client could request applications "foo" and "bar" from the API, and the User Agent would send ALPNs "wq:foo" and "wq:bar" in its client handshake as an offer.  The server would have to pick one of them or abort the connection.

**Comment by pthatcherg** [comment]
Seems like an interesting way to piggyback some info on the QUIC handshake.  A web app could always do the same with messages/streams/datagrams after the QUIC handshake completes, but this is just slightly faster.    So like all optimizations, the question is: is it worth it?  

To know if it's worth it, it would be nice to know what kinds of subprotocols would be used.  Basically, what use case would be faster that would make this extra complexity worth it?

**Comment by vasilvv** [comment]
I agree that we should think hard about what exactly do we need here any why.  I am not entirely clear on this myself, but I will note the following consideration.

Both HTTP and WSP allow the client to specify sub-endpoint within the server host (as /path).  This is convenient in deployments for composing reasons: you can have two servers maintained by different teams running as standalone processes, and you can then multiplex them using a reverse proxy (e.g. nginx), dispatching by path or subprotocol.  This does require a standardized way to specify sub-endpoint.

**Comment by pthatcherg** [comment]
My understanding of the use case:  some JS wants to say QuicTransport("host", port, "chat") and another wants to say QuicTransport("host", port, "notifications").  It's the same host, but different services.  Where do we put those services on the wire?  The app *could* put that in a stream of its own, but that adds some app-level complexity and an RTT.

**Comment by pthatcherg** [comment]
One issue with it being in the ALPN is that it won't be encrypted,  which might be a surprise and problem for the app.

**Comment by pthatcherg** [comment]
One alternative might be to do what we've always wanted to avoid: have QuicTransport allocate stream 1 and use that for sending sub protocols (and possibly other stuff).  Maybe we should just reserve stream 1 for future extensiblity.

**Comment by pthatcherg** [comment]
If it's not encrypted, and we're OK with that, we might as well use QUIC transport parameters, right?

**Comment by aboba** [comment]
Need to add optional third parameter to the RTCQuicTransport constructor.
